src/openvic-simulation/economy/production/ArtisanalProducer.cpp

```diff
@@ -1,150 +1,157 @@
+			fixed_point_t total_stockpile_value = fixed_point_t::_0();
+			for (auto const& [input_good_ptr, max_price] : goods_to_buy_and_max_price) {
+				total_demand_value += max_price * demand[input_good_ptr];
+				total_stockpile_value += max_price * stockpile[input_good_ptr];
+			}
 
-				if ( total_demand_value == fixed_point_t::_0()) {
+			if ( total_demand_value == fixed_point_t::_0()) {
+				max_possible_satisfaction_numerator = fixed_point_t::_1();
+				max_possible_satisfaction_denominator = fixed_point_t::_1();
+			} else {
+				max_possible_satisfaction_numerator = total_stockpile_value + total_cash_to_spend;
+				max_possible_satisfaction_denominator = total_demand_value;
+				if (max_possible_satisfaction_numerator > max_possible_satisfaction_denominator) {
 					max_possible_satisfaction_numerator = fixed_point_t::_1();
 					max_possible_satisfaction_denominator = fixed_point_t::_1();
-				} else {
-					max_possible_satisfaction_numerator = total_stockpile_value + total_cash_to_spend;
-					max_possible_satisfaction_denominator = total_demand_value;
-					if (max_possible_satisfaction_numerator > max_possible_satisfaction_denominator) {
-						max_possible_satisfaction_numerator = fixed_point_t::_1();
-						max_possible_satisfaction_denominator = fixed_point_t::_1();
-					}
-				}
-
-				for (auto const& [input_good_ptr, max_price] : goods_to_buy_and_max_price) {
-					const fixed_point_t optimal_quantity = demand[input_good_ptr] * max_possible_satisfaction_numerator / max_possible_satisfaction_denominator;
-					if (stockpile[input_good_ptr] >= optimal_quantity) {
-						goods_to_buy_and_max_price.erase(input_good_ptr);
-						at_or_below_optimum = false;
-					}
 				}
 			}
 
-			//Place buy orders for each input
 			for (auto const& [input_good_ptr, max_price] : goods_to_buy_and_max_price) {
-				const fixed_point_t good_demand = demand[input_good_ptr];
-				fixed_point_t& good_stockpile = stockpile[input_good_ptr];
-				const fixed_point_t optimal_quantity = good_demand * max_possible_satisfaction_numerator / max_possible_satisfaction_denominator;
-				const fixed_point_t max_quantity_to_buy = good_demand - good_stockpile;
-				if (max_quantity_to_buy > 0) {
-					const fixed_point_t money_to_spend = optimal_quantity * max_price;
-					market_instance.place_buy_up_to_order({
-						*input_good_ptr,
-						max_quantity_to_buy,
-						money_to_spend,
-						[this, &pop, &good_stockpile](const BuyResult buy_result) -> void {
-							if (buy_result.get_quantity_bought() > fixed_point_t::_0()) {
-								pop.add_artisan_inputs_expense(buy_result.get_money_spent());
-								good_stockpile += buy_result.get_quantity_bought();
-							}
-						}
-					});
+				const fixed_point_t optimal_quantity = demand[input_good_ptr] * max_possible_satisfaction_numerator / max_possible_satisfaction_denominator;
+				if (stockpile[input_good_ptr] >= optimal_quantity) {
+					goods_to_buy_and_max_price.erase(input_good_ptr);
```

Comment:
I'm not sure if it's safe to erase map elements while iterating through them 😕

Reply:
Nice catch. A similar piece of code works fine as the loop uses a different collection.
Here it would probably screw things up.

Reply:
There's still a risk that the iterators get invalidated here, in general you can't trust any iterators to remain valid after adding or removing an element from a container unless the container specifically guarantees it. The safe approach is to use `erase_if`:
```suggestion
			erase_if(
				goods_to_buy_and_max_price,
				[
					this, &demand, max_possible_satisfaction_numerator, max_possible_satisfaction_denominator,
					&at_or_below_optimum
				](auto const& pair) {
					GoodDefinition const* const input_good_ptr = pair.first;
					const fixed_point_t optimal_quantity =
						demand[input_good_ptr] * max_possible_satisfaction_numerator / max_possible_satisfaction_denominator;
					if (stockpile[input_good_ptr] >= optimal_quantity) {
						at_or_below_optimum = false;
						return true;
					}
					return false;
				}
```